### PR TITLE
Alerting: Decouple rule routine from scheduler

### PR DESCRIPTION
**Pull request by alexweav**
**What is this feature?**

The prior PR made the ruleRoutine take a scheduler parameter to handle its schedule dependencies.

Now, we remove this, and put the relevant dependencies directly on the rule. The rule contains everything it needs to run independently of the scheduler now.

None of these are stateful, and they're all reference types aside from a couple configuration options.
We rename `ruleRoutine` to `run` to match the rest of the rule process API that we're building.

**Which issue(s) does this PR fix?**:

n/a

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/ngalert/schedule/alert_rule.go [hunk 1]
```diff
@@ -18,20 +23,114 @@ import (
...
+
+	appURL               *url.URL
+	disableGrafanaFolder bool
+	maxAttempts          int64
+
+	clock        clock.Clock
+	sender       AlertsSender
+	stateManager *state.Manager
+	evalFactory  eval.EvaluatorFactory
+	ruleProvider ruleProvider
```
Comment (alexweav) [inline pkg/services/ngalert/schedule/alert_rule.go:88]: Many of these are duplicated from the scheduler. Several are no longer used directly by the scheduler and are only used by the rule. We'll simplify this in a future PR.

File: pkg/services/ngalert/schedule/alert_rule.go [hunk 2]
```diff
@@ -82,52 +181,52 @@ func (a *alertRuleInfo) stop(reason error) {
...
 
 	evaluate := func(ctx context.Context, f fingerprint, attempt int64, e *evaluation, span trace.Span, retry bool) error {
 		logger := logger.New("version", e.rule.Version, "fingerprint", f, "attempt", attempt, "now", e.scheduledAt).FromContext(ctx)
-		start := sch.clock.Now()
+		start := a.clock.Now()
 
-		evalCtx := eval.NewContextWithPreviousResults(ctx, SchedulerUserFor(e.rule.OrgID), sch.newLoadedMetricsReader(e.rule))
-		if sch.evaluatorFactory == nil {
+		evalCtx := eval.NewContextWithPreviousResults(ctx, SchedulerUserFor(e.rule.OrgID), a.newLoadedMetricsReader(e.rule))
+		if a.evalFactory == nil {
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/schedule/alert_rule.go:218]: Let's make a todo to fix this. I think we should return an error in `newRuleFactory` if this is nil.

**Review by grobinson-grafana — APPROVED** [review]

File: pkg/services/ngalert/schedule/alert_rule.go (hunk 2 above)
Comment (alexweav) [reply pkg/services/ngalert/schedule/alert_rule.go:218]: heh... this is a testing statement I used in a PR 2 months ago and forgot to remove :laughing:

Comment (alexweav) [reply pkg/services/ngalert/schedule/alert_rule.go]: i just removed it directly

**Pull request merged**